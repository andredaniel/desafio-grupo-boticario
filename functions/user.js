const axios = require('axios')

const randomStatus = () => {
  const key = Math.floor(Math.random() * 3)

  const statuses = ['APPROVED', 'DISAPPROVED', 'IN_VALIDATION']

  return statuses[key]
}

exports.handler = async function (event, context) {
  const baseUrl = 'https://5fffb468cb21e10017af7e99.mockapi.io/api'

  if (event.httpMethod === 'GET') {
    const { data } = await axios.get(
      `${baseUrl}/orders/?user_id=${event.queryStringParameters.user_id}`
    )

    return {
      statusCode: 200,
      body: JSON.stringify(data),
    }
  }

  if (event.httpMethod === 'POST') {
    const newOrder = JSON.parse(event.body)

    newOrder.cashbackAppliedPercentage = Math.ceil(Math.random() * 20)
    newOrder.cashbackReceivedValue =
      (newOrder.value / 100) * newOrder.cashbackAppliedPercentage
    newOrder.status = randomStatus()

    const { data } = await axios.post(`${baseUrl}/orders`, newOrder)

    return {
      statusCode: 200,
      body: JSON.stringify(data),
    }
  }
}
