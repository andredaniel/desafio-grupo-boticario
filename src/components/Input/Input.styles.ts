import styled from 'styled-components'
import { theme } from '../../theme'

export const InputWrapper = styled.div`
  padding: 0.5rem 0;
  width: 100%;

  label {
    display: block;
    margin-bottom: 5px;
  }
  
  input {
    background: #fff;
    border-radius: 0.4rem;
    border: 1px solid #ddd;
    padding: 0.7rem 1rem;
    width: 100%;
    font-size: 1rem;

    &:focus {
      outline: none;
      border-color: ${theme.colors.primary.MAIN};
    }
  }
`
