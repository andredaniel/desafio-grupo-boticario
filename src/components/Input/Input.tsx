import React, { InputHTMLAttributes } from 'react'
import InputMask from 'react-input-mask'
import { InputWrapper } from './Input.styles'

interface IInput extends InputHTMLAttributes<HTMLInputElement> {
  label?: string
  mask?: string
}

const Input: React.FC<IInput> = (props): JSX.Element => {
  const { label, id, type, mask = '', ...rest } = props

  return (
    <InputWrapper>
      {label && <label htmlFor={id}>{label}</label>}
      <InputMask
        id={id}
        mask={mask}
        maskPlaceholder={null}
        type={type || 'text'}
        {...rest}
      />
    </InputWrapper>
  )
}

export { Input }
export type { IInput }
