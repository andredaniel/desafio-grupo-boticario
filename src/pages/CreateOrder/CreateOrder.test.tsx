import { render } from '@testing-library/react'
import React from 'react'
import { BrowserRouter as Router } from 'react-router-dom'
import { changeInputMaskValue } from '../../components/Input/Input.test'
import { UserProvider } from '../../contexts/user.context'
import Page from './CreateOrder'

describe('<Page />', () => {
  it('should render Page component', () => {
    const { container } = render(
      <UserProvider>
        <Router>
          <Page />
        </Router>
      </UserProvider>
    )

    expect(container).toMatchSnapshot()
  })

  it('should fill the form and create a new order', async () => {
    const { container, getByPlaceholderText } = render(
      <UserProvider>
        <Router>
          <Page />
        </Router>
      </UserProvider>
    )

    const inputCode = getByPlaceholderText(/CÃ³digo da compra/i)
    const inputValue = getByPlaceholderText(/Valor da compra/i)
    const inputDate = getByPlaceholderText(/Data da compra/i)

    changeInputMaskValue(inputCode, '12345')
    changeInputMaskValue(inputValue, 100)
    changeInputMaskValue(inputDate, '28/09/2021')

    expect(container).toMatchSnapshot()
  })
})
