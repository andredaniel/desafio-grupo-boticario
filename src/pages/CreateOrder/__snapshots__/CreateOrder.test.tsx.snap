// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> should render an Input without mask 1`] = `
<div>
  <div
    class="sc-bdfBwQ jSMxIb"
  >
    <input
      type="text"
    />
  </div>
</div>
`;

exports[`<Page /> should fill the form and create a new order 1`] = `
<div>
  <div
    class="sc-kEjbxe iXipqX"
  >
    <div
      class="sc-jrAGrp ldoQrE"
      y="0"
    >
      <div
        class="brand-wrapper"
      >
        <svg>
          logo-black.svg
        </svg>
        <span>
          Cashback
        </span>
      </div>
      <div
        class="user-info"
      >
        <img
          class="sc-gsTCUz cMuDZN"
          src=""
        />
        <span
          class="welcome-message"
        >
          Ol치, 
          !
        </span>
      </div>
    </div>
    <div
      class="sc-iBPRYJ jtyoTZ"
    >
      <div
        class="sc-fubCfw cnDguJ"
      >
        <div
          class="sc-pFZIQ hYJJFY"
        >
          <a
            aria-current="page"
            class="active"
            href="/"
          >
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M3 9l9-7 9 7v11a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2z"
              />
              <polyline
                points="9 22 9 12 15 12 15 22"
              />
            </svg>
            <span>
              Dashboard
            </span>
            <canvas
              height="0"
              style="border-radius: inherit; height: 100%; left: 0px; position: absolute; top: 0px; width: 100%;"
              width="0"
            />
          </a>
        </div>
        <div
          class="sc-pFZIQ hYJJFY"
        >
          <a
            href="/orders/create"
          >
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="18"
                rx="2"
                ry="2"
                width="18"
                x="3"
                y="3"
              />
              <line
                x1="12"
                x2="12"
                y1="8"
                y2="16"
              />
              <line
                x1="8"
                x2="16"
                y1="12"
                y2="12"
              />
            </svg>
            <span>
              Nova compra
            </span>
            <canvas
              height="0"
              style="border-radius: inherit; height: 100%; left: 0px; position: absolute; top: 0px; width: 100%;"
              width="0"
            />
          </a>
        </div>
        <div
          class="sc-pFZIQ hYJJFY"
        >
          <a
            href="/orders"
          >
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M6 2L3 6v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V6l-3-4z"
              />
              <line
                x1="3"
                x2="21"
                y1="6"
                y2="6"
              />
              <path
                d="M16 10a4 4 0 0 1-8 0"
              />
            </svg>
            <span>
              Minhas compras
            </span>
            <canvas
              height="0"
              style="border-radius: inherit; height: 100%; left: 0px; position: absolute; top: 0px; width: 100%;"
              width="0"
            />
          </a>
        </div>
        <div
          class="sc-pFZIQ hYJJFY"
        >
          <a
            href="/logout"
          >
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M9 21H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h4"
              />
              <polyline
                points="16 17 21 12 16 7"
              />
              <line
                x1="21"
                x2="9"
                y1="12"
                y2="12"
              />
            </svg>
            <span>
              Sair
            </span>
            <canvas
              height="0"
              style="border-radius: inherit; height: 100%; left: 0px; position: absolute; top: 0px; width: 100%;"
              width="0"
            />
          </a>
        </div>
      </div>
    </div>
    <div
      class="sc-iqHYGH bmLHrU"
    >
      <h1
        class="title"
      >
        Cadastrar nova compra
      </h1>
      <form
        action="#"
      >
        <div
          class="sc-gKsewC kBRlih"
        >
          <div
            class="sc-bdfBwQ jSMxIb"
          >
            <label
              for="code"
            >
              C칩digo
            </label>
            <input
              class=""
              id="code"
              name="code"
              placeholder="C칩digo da compra"
              type="text"
              value="12345"
            />
          </div>
        </div>
        <div
          class="sc-gKsewC kBRlih"
        >
          <div
            class="sc-bdfBwQ jSMxIb"
          >
            <label
              for="value"
            >
              Valor
            </label>
            <input
              class=""
              id="value"
              name="value"
              placeholder="Valor da compra"
              type="text"
              value="R$ ___,__"
            />
          </div>
        </div>
        <div
          class="sc-gKsewC kBRlih"
        >
          <div
            class="sc-bdfBwQ jSMxIb"
          >
            <label
              for="date"
            >
              Data
            </label>
            <input
              class=""
              id="date"
              name="date"
              placeholder="Data da compra"
              type="text"
              value="28/09/2021"
            />
          </div>
        </div>
        <div
          class="sc-dlfnbm iVvQwa"
        >
          <button
            type="submit"
          >
            Cadastrar compra
            <canvas
              height="0"
              style="border-radius: inherit; height: 100%; left: 0px; position: absolute; top: 0px; width: 100%;"
              width="0"
            />
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`<Page /> should render Page component 1`] = `
<div>
  <div
    class="sc-kEjbxe iXipqX"
  >
    <div
      class="sc-jrAGrp ldoQrE"
      y="0"
    >
      <div
        class="brand-wrapper"
      >
        <svg>
          logo-black.svg
        </svg>
        <span>
          Cashback
        </span>
      </div>
      <div
        class="user-info"
      >
        <img
          class="sc-gsTCUz cMuDZN"
          src=""
        />
        <span
          class="welcome-message"
        >
          Ol치, 
          !
        </span>
      </div>
    </div>
    <div
      class="sc-iBPRYJ jtyoTZ"
    >
      <div
        class="sc-fubCfw cnDguJ"
      >
        <div
          class="sc-pFZIQ hYJJFY"
        >
          <a
            aria-current="page"
            class="active"
            href="/"
          >
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M3 9l9-7 9 7v11a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2z"
              />
              <polyline
                points="9 22 9 12 15 12 15 22"
              />
            </svg>
            <span>
              Dashboard
            </span>
            <canvas
              height="0"
              style="border-radius: inherit; height: 100%; left: 0px; position: absolute; top: 0px; width: 100%;"
              width="0"
            />
          </a>
        </div>
        <div
          class="sc-pFZIQ hYJJFY"
        >
          <a
            href="/orders/create"
          >
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="18"
                rx="2"
                ry="2"
                width="18"
                x="3"
                y="3"
              />
              <line
                x1="12"
                x2="12"
                y1="8"
                y2="16"
              />
              <line
                x1="8"
                x2="16"
                y1="12"
                y2="12"
              />
            </svg>
            <span>
              Nova compra
            </span>
            <canvas
              height="0"
              style="border-radius: inherit; height: 100%; left: 0px; position: absolute; top: 0px; width: 100%;"
              width="0"
            />
          </a>
        </div>
        <div
          class="sc-pFZIQ hYJJFY"
        >
          <a
            href="/orders"
          >
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M6 2L3 6v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V6l-3-4z"
              />
              <line
                x1="3"
                x2="21"
                y1="6"
                y2="6"
              />
              <path
                d="M16 10a4 4 0 0 1-8 0"
              />
            </svg>
            <span>
              Minhas compras
            </span>
            <canvas
              height="0"
              style="border-radius: inherit; height: 100%; left: 0px; position: absolute; top: 0px; width: 100%;"
              width="0"
            />
          </a>
        </div>
        <div
          class="sc-pFZIQ hYJJFY"
        >
          <a
            href="/logout"
          >
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M9 21H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h4"
              />
              <polyline
                points="16 17 21 12 16 7"
              />
              <line
                x1="21"
                x2="9"
                y1="12"
                y2="12"
              />
            </svg>
            <span>
              Sair
            </span>
            <canvas
              height="0"
              style="border-radius: inherit; height: 100%; left: 0px; position: absolute; top: 0px; width: 100%;"
              width="0"
            />
          </a>
        </div>
      </div>
    </div>
    <div
      class="sc-iqHYGH bmLHrU"
    >
      <h1
        class="title"
      >
        Cadastrar nova compra
      </h1>
      <form
        action="#"
      >
        <div
          class="sc-gKsewC kBRlih"
        >
          <div
            class="sc-bdfBwQ jSMxIb"
          >
            <label
              for="code"
            >
              C칩digo
            </label>
            <input
              class=""
              id="code"
              name="code"
              placeholder="C칩digo da compra"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="sc-gKsewC kBRlih"
        >
          <div
            class="sc-bdfBwQ jSMxIb"
          >
            <label
              for="value"
            >
              Valor
            </label>
            <input
              class=""
              id="value"
              name="value"
              placeholder="Valor da compra"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="sc-gKsewC kBRlih"
        >
          <div
            class="sc-bdfBwQ jSMxIb"
          >
            <label
              for="date"
            >
              Data
            </label>
            <input
              class=""
              id="date"
              name="date"
              placeholder="Data da compra"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="sc-dlfnbm iVvQwa"
        >
          <button
            type="submit"
          >
            Cadastrar compra
            <canvas
              height="0"
              style="border-radius: inherit; height: 100%; left: 0px; position: absolute; top: 0px; width: 100%;"
              width="0"
            />
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
